### ISSUE TITLE
Internal Server Error when previewing hidden pages

### ISSUE THREAD

**Issue by kimchristiansen** [FluidTYPO3/flux#1618]
If i want to preview a hidden page, i get an internal server error. (apache log: timeout, incomplete headers)
The error appears in all hidden pages, also in pages with no content elements.

Configuration:
flux 9.0.1 
fluidpages 4.2.0 
TYPO3 8.7.19

The error appeared after updating flux from 8.x to 9.0.1
Same error in TYPO3 8.7.17 and 8.7.18 and with flux 9.0.0

PHP Log says: Maximum execution time of 512 seconds exceeded in /typo3conf/ext/flux/Classes/Service/RecordService.php on line 73

Does anybody have the same issue?

**Comment by kimchristiansen** [comment]
The error still occures in flux 9.1.0 and fluidpages 4.3.0 with TYPO3 8.7.19

**Comment by glucka** [comment]
#1618 should solve it

**Comment by thopra** [comment]
Since the PR above has been closed, does anyone have a sensible workaround for this? Right now we can't preview hidden pages, which is kind of a blocker.

**Comment by thopra** [comment]
Unfortunatly, removing fluidpages would require some refactoring.. 
I ended up XClassing the Service causing the issue instead:

```
<?php

namespace Vendor\Extension\Xclass\Flux\Service;

use TYPO3\CMS\Core\SingletonInterface;
use TYPO3\CMS\Core\Database\Query\Restriction\HiddenRestriction;
use TYPO3\CMS\Core\Utility\ExtensionManagementUtility;

/**
 * XCLASS for \FluidTYPO3\Flux\Service\RecordService as a temporary workaround to "fix" https://github.com/FluidTYPO3/flux/issues/1618
 *
 * @todo remove when the issue is resolved and flux is updated
 */
class WorkspacesAwareRecordService extends \FluidTYPO3\Flux\Service\WorkspacesAwareRecordService implements SingletonInterface
{
    /**
     * @param string $table
     * @param string $fields
     * @param string $uid
     * @return array|NULL
     */
    public function getSingle($table, $fields, $uid)
    {
        if (version_compare(ExtensionManagementUtility::getExtensionVersion('flux'), '9.1.0') > 0) {
            throw new \Exception('Flux has been updated. Check if this XClass ('.self::class.') is outdated and https://github.com/FluidTYPO3/flux/issues/1618 has been resolved!');
        }
        $record = parent::getSingle($table, $fields, $uid);

        if (TYPO3_MODE !== 'BE' && !$record && $table == 'pages' && $GLOBALS['TSFE']->fePreview) {
            $record = $this->getSingleWithoutRestriction($table, $fields, $uid);
        }

        if ($record) {
            $overlay = $this->overlayRecord($table, $record);
            if ($overlay) {
                return $overlay;
            }
        }
        return $record;
    }

    public function getSingleWithoutRestriction($table, $fields, $uid)
    {
        $q = $this->getQueryBuilder($table);
        $q->getRestrictions()->removeByType(HiddenRestriction::class);
        $results = $q
            ->from($table)
            ->select(...explode(',', $fields))
            ->where(sprintf('uid = %d', $uid))
            ->execute()
            ->fetchAll() ?: [];
        return reset($results);
    }

}

```

This solves the specific case (a hidden page that inherits the fluidpages template is being previewed). Not sure about side effects yet though - and not tested with workspaces (but who uses these anyways :)).

**Comment by NamelessCoder** [comment]
```
        if (version_compare(ExtensionManagementUtility::getExtensionVersion('flux'), '9.1.0') > 0) {
            throw new \Exception('Flux has been updated. Check if this XClass ('.self::class.') is outdated and https://github.com/FluidTYPO3/flux/issues/1618 has been resolved!');
        }
```

@thopra That's an interesting, albeit extremely aggressive, solution to inform yourself about when to drop the XCLASS! :)

**Comment by thopra** [comment]
heh, ikr - tend to forget about them which caused me some headache in the past...

**Comment by foobar13372** [comment]
What is the state here? Out customers can't use the preview, which is a severe blocker for them.

**Comment by felixbuenemann** [comment]
@NamelessCoder What do you think about fixing this by inspecting the preview settings from the frontend controller?

The following patch works for me, but I'm not sure if it's the right way to go:

```diff
diff --git a/Classes/Service/RecordService.php b/Classes/Service/RecordService.php
index 10e76ee7..874ed2f3 100644
--- a/Classes/Service/RecordService.php
+++ b/Classes/Service/RecordService.php
@@ -11,6 +11,9 @@ namespace FluidTYPO3\Flux\Service;
 use TYPO3\CMS\Backend\Utility\BackendUtility;
 use TYPO3\CMS\Core\Database\ConnectionPool;
 use TYPO3\CMS\Core\Database\Query\QueryBuilder;
+use TYPO3\CMS\Core\Database\Query\Restriction\EndTimeRestriction;
+use TYPO3\CMS\Core\Database\Query\Restriction\HiddenRestriction;
+use TYPO3\CMS\Core\Database\Query\Restriction\StartTimeRestriction;
 use TYPO3\CMS\Core\SingletonInterface;
 use TYPO3\CMS\Core\Utility\GeneralUtility;

@@ -118,7 +121,20 @@ class RecordService implements SingletonInterface
      */
     protected function getQueryBuilder($table)
     {
-        return GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable($table);
+        $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable($table);
+        $includeHidden = $GLOBALS['TSFE']->fePreview && (
+            'pages' === $table
+                ? $GLOBALS['TSFE']->showHiddenPage
+                : $GLOBALS['TSFE']->showHiddenRecords
+        );
+        if ($includeHidden) {
+            $queryBuilder
+                ->getRestrictions()
+                ->removeByType(HiddenRestriction::class)
+                ->removeByType(StartTimeRestriction::class)
+                ->removeByType(EndTimeRestriction::class);
+        }
+        return $queryBuilder;
     }

 }
```

**Comment by NamelessCoder** [comment]
@felixbuenemann It's quite risky to change this method because it's used without discretion in frontend and backend (except for some cases where greater control was required). Generally speaking I'd prefer a solution that avoids this service in the cases where this is possible. If there is core API to be used - e.g. methods on Connection which returns a pre-configured QueryBuilder that takes these things into consideration - I'd much rather use that. I should mention that RecordService comes from a time when we supported legacy DB and Doctrine through a single class and now that we're fully using Doctrine, there's not as much merit for a service to manage the differences in API.

In this case I think we should switch the place(s) that require previewing records so they use QueryBuilder. We should also look for possible API (on Connection?) which returns the right restrictions so we avoid reproducing the various conditions in our own code.

**Comment by felixbuenemann** [comment]
Most of the logic I found for this is contained in the TYPO3 Frontend-Controller and duplicated in some places for things like Frontend Editing. I didn't find a common public api to do it, though it's entirely possible I missed it while searching…

**Comment by glucka** [comment]
This can be closed? See 482dd13510688791d1cd551f68f35f6a6dde9721

**Comment by kimchristiansen** [comment]
After updating fluidpages to the development branch, i get the following error on hidden pages:
`#1257246929: Tried resolving a template file for controller action "Page->default" in format ".html", but none of the paths contained the expected template file (Page/Default.html). The following paths were checked: [projectpath]/typo3conf/ext/fluidpages/Resources/Private/Templates/`
The error only appears on hidden pages.
